frontend/src/features/contacts/contact/create/Person/CreatePersonForm.tsx

```diff
@@ -0,0 +1,262 @@
+      // combine yup schema validation with custom rules
+      const errors = {} as any;
+      if (!hasEmail(values) && !hasPhoneNumber(values) && !hasAddress(values)) {
+        errors.needsContactMethod =
+          'Contacts must have a minimum of one method of contact to be saved. (ex: email,phone or address)';
+      }
+      return errors;
+    } catch (err) {
+      return yupToFormErrors(err);
+    }
+  };
+
+  const onSubmit = async (
+    values: ICreatePersonForm,
+    { resetForm, setSubmitting }: FormikHelpers<ICreatePersonForm>,
+  ) => {
+    try {
+      await postPerson(personCreateFormToApiPerson(values));
+      setSubmitting(false);
+      resetForm({ values });
+      toast.info('Contact added successfully');
+      history.push('/contact/list');
+    } catch (error) {
+      toast.error('Failed to add contact', { autoClose: 7000 });
+      history.push('/contact/list');
+    }
+  };
+
+  return (
+    <Formik
+      component={CreatePersonComponent}
+      initialValues={defaultCreatePerson}
+      enableReinitialize
+      validate={onValidate}
+      onSubmit={onSubmit}
+    />
+  );
+};
+
+export default CreatePersonForm;
+
+/**
+ * Sub-component that is wrapped by Formik
+ */
+const CreatePersonComponent: React.FC<FormikProps<ICreatePersonForm>> = ({
+  values,
+  errors,
+  touched,
+  dirty,
+  resetForm,
+  submitForm,
+  initialValues,
+}) => {
+  const history = useHistory();
+  const [showConfirmation, setShowConfirmation] = useState(false);
+
+  // organization type-ahead state
+  const { getOrganizationPredictions } = useApiAutocomplete();
+  const [isTypeaheadLoading, setIsTypeaheadLoading] = useState(false);
+  const [matchedOrgs, setMatchedOrgs] = useState<IAutocompletePrediction[]>([]);
+
+  // fetch autocomplete suggestions from server
+  const handleTypeaheadSearch = async (query: string) => {
+    try {
+      setIsTypeaheadLoading(true);
+      const { data } = await getOrganizationPredictions(query);
+      setMatchedOrgs(data.predictions);
+      setIsTypeaheadLoading(false);
+    } catch (e) {
+      setMatchedOrgs([]);
+      toast.error('Failed to get autocomplete results for supplied organization', {
+        autoClose: 7000,
+      });
+    } finally {
+      setIsTypeaheadLoading(false);
+    }
+  };
+
+  const onCancel = () => {
+    if (dirty) {
+      setShowConfirmation(true);
+    } else {
+      history.push('/contact/list');
+    }
+  };
+
+  return (
+    <>
+      {/* Router-based onfirmation popup when user tries to navigate away and form has unsaved changes */}
+      <UnsavedChangesPrompt />
+
+      {/* Confirmation popup when Cancel button is clicked */}
+      <GenericModal
+        title="Unsaved Changes"
+        message="Confirm cancel adding this contact? Changes will not be saved."
+        cancelButtonText="No"
+        okButtonText="Confirm"
+        display={showConfirmation}
+        handleOk={() => {
```

Comment:
strange, they want the route guard and a navigate away warning?

Reply:
As per business requirements:

```
14. Cancel button* - on click:
If there is data in any fields, show modal popup: "Confirm cancel adding this contact? Changes will not be saved. [No] [Confirm]"
*No - remove modal, allow user to continue editing
*Confirm - do not save the record, and return the user to the Contact List screen.

If there is no data in any fields, do not save the record, and return the user to the Contact List screen.

16. If the form is partially filled, and the user clicks the other button, it would be advised to show the modal related to navigate away from the page "If the screen is closed before saving these changes, they will be lost. Do you want to continue without saving?"
```

Reply:
In interpret the above as TWO separate popups. What do you think?

Reply:
As I see a specific rules for the Cancel button. Requirement #16 speaks about when you click any other link in the app that causes it to navigate away (at least that is how I saw it)

Reply:
I think this is correct given the above, but doesn't match the requirements I've been given on other forms. sigh.

Reply:
We will need to take it back to the business to define a common approach to Cancel & navigate away for all forms